app/models/enterprise.rb

```diff
@@ -220,6 +220,22 @@ def relatives
     ", self.id, self.id)
   end
 
+  def relatives_and_oc_producers(order_cycles)
+    enterprise_ids = []
+    order_cycles.each do |oc|
+      enterprise_ids += oc.exchanges.incoming.pluck :sender_id
+    end
```

Comment:
In terms of SQL queries, this section is O(n) in the number of order cycles. We could make it O(1) like this:

```ruby
Exchange.in_order_cycle(order_cycles).incoming.pluck(:sender_id)
```